spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java

```diff
@@ -159,16 +226,187 @@ public ListenableFuture<Job> writeDataToTable(
   }
 
   /**
-   * @return the name of the BigQuery dataset that the template is operating in.
+   * This method uses BigQuery Storage Write API to write new line delimited JSON file to the
+   * specified table. This method creates a table with the specified schema.
+   *
+   * @param tableName name of the table to write to
+   * @param jsonInputStream input stream of the json file to be written
+   * @return {@link ListenableFuture} containing the WriteApiResponse indicating completion of
+   *     operation
+   */
+  @Override
+  public ListenableFuture<WriteApiResponse> writeJsonStream(
+      String tableName, InputStream jsonInputStream, Schema schema) {
+    createTable(tableName, schema); // create table if it's not already created
+    return writeJsonStream(tableName, jsonInputStream);
+  }
+
+  @VisibleForTesting
+  public Table createTable(
+      String tableName, Schema schema) { // create table if it's not already created
+    TableId tableId = TableId.of(datasetName, tableName);
+    Table table = bigQuery.getTable(TableId.of(datasetName, tableName));
+    if (table == null || !table.exists()) {
+      TableDefinition tableDefinition = StandardTableDefinition.of(schema);
+      TableInfo tableInfo = TableInfo.newBuilder(tableId, tableDefinition).build();
+      return bigQuery.create(tableInfo);
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * This method uses BigQuery Storage Write API to write new line delimited JSON file to the
+   * specified table. The Table should already be created as BigQuery Storage Write API doesn't
+   * create it automatically.
+   *
+   * @param tableName name of the table to write to
+   * @param jsonInputStream input stream of the json file to be written
+   * @return {@link ListenableFuture} containing the WriteApiResponse indicating completion of
+   *     operation
    */
+  @Override
+  public ListenableFuture<WriteApiResponse> writeJsonStream(
+      String tableName, InputStream jsonInputStream) {
+
+    SettableListenableFuture<WriteApiResponse> writeApiFutureResponse =
+        new SettableListenableFuture<>();
+
+    Thread asyncTask =
+        new Thread(
```

Comment:
It seems dangerous to create an unbounded number of threads without a thread pool here. It might be causing the OutOfMemoryError reported in #1599.

cc: @prash-mi, @burkedavison

Reply:
+1 : Stack trace in the issue points to [asyncTask.start()](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/v3.4.0/spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java#L310)

Good stackoverflow on the issue here: https://stackoverflow.com/a/16789621